### DISCUSSION TITLE
Code Review Problems

### DISCUSSION THREAD

**Discussion by The-Futurist** [community/community#23331] [New to GitHub]
<p>I’m having trouble pinning down the exact steps for us to do this correctly.</p>
<p>We want to have all PRs from a <em>project_branch</em> in people’s forks to the same <em>project_branch</em> in the main repo (these are private organization repos) to be forcibly subject to reviews.</p>
<p>In order to facilitate this I added a branch protection rule to all project branches in the main repo and in that rule specified we need 2 approvals before merging is allowed.</p>
<p>However when we do this we get a problem, imagine we’re now done with <em>project_branch</em> in the main repo so we create a PR to merge that into master. Well when we do this GitHub ordinarily tells us we can now delete <em>project_branch</em> but now it does not and in fact it tells us that the branch cannot be deleted!</p>
<p>It seems that it isn’t possible to apply this two approval rule to project branches without also making the branch not deleteable but this seems dumb.</p>
<p>Furthermore we see weird stuff when we approve, we expected (after two approvals) that GitHub would automatically merge the PR but it didn’t and in fact at the top of the PR page (top right) we see clearly that three people approved the PR yet at the bottom of the same page it says only a single person has approved.</p>
<p>This is just driving us insane trying to fathom this out.</p>
<p>Thx</p>

**Comment by nadiajoyce** [comment]
<p>Hi <a class="mention" href="/korporal">@korporal</a>,</p>
<p>Thanks for being part of the GitHub Community Forum! I’ll do all I can to help.</p>
<p>To address your question about not being able to delete that branch, this is expected behavior per the <a href="https://help.github.com/articles/about-protected-branches/#branch-protection-settings" rel="nofollow noopener">documentation on branch protections</a>: </p>
<p>&gt; When you create a protected branch rule in a repository, collaborators cannot force push to the protected branch or delete the branch by default.</p>
<p>That said, a repo <em>admin</em> should still be able to delete that branch (unless, when setting up the branch protections, “Include administrators” was checked). If you want repo admins to be able to delete branches (or otherwise circumvent the branch protections), you’ll want to make sure that checkbox is left unchecked.</p>
<p>As for some of the other things you describe:</p>
<p>&gt;…we expected (after two approvals) that GitHub would automatically merge the PR…</p>
<p>GitHub does not automatically merge PRs this way. A protected branch with required approvals will be unmergable until the approvals have occurred; however, collecting the required approvals only makes the PR mergable - it does not merge the PR automatically. This is expected behavior.</p>
<p>&gt;…at the top of the PR page (top right) we see clearly that three people approved the PR yet at the bottom of the same page it says only a single person has approved…</p>
<p>This could be for a couple of reasons. If anyone outside of your organization approves the PR, those reviews wouldn’t count as “approvals” for the count at the bottom. Also, if you’re using <code>CODEOWNERS</code> or something else that restricts who can actually <em>approve</em> merges, the count at the bottom will also show a different number if just a member of the org (but not a <code>CODEOWNER</code>) approves the PR.</p>
<p>I hope this helps. Please let us know if you have more questions!</p>

**Comment by Grabauskas** [comment]
<blockquote>
<p>That said, a repo <em>admin</em> should still be able to delete that branch (unless, when setting up the branch protections, “Include administrators” was checked). If you want repo admins to be able to delete branches (or otherwise circumvent the branch protections), you’ll want to make sure that checkbox is left unchecked.</p>
</blockquote>
<p>It doesn’t work, still having the same rejection.</p>
<p>GitHub UI:</p>
<pre><code>You can't delete a protected branch.
</code></pre>
<p>Force push:</p>
<pre><code>remote: error: GH006: Protected branch update failed for refs/heads/**.
remote: error: Cannot delete a protected branch
To https://github.com/ **/**.git
 ! [remote rejected] ** (protected branch hook declined)
error: failed to push some refs to 'https://github.com/ **/**.git'
</code></pre>
<p>&gt; …<em>enforce certain workflows or requirements…</em></p>
<p>How I understand, that certain workflows or requirements only available via pull requests. It’s that correct, <a class="mention" href="/nadiajoyce">@nadiajoyce</a>?</p>

**Comment by nadiajoyce** [comment]
<p><a class="mention" href="/grabauskas">@grabauskas</a> Are you certain that “Include administrators” isn’t checked when setting up branch protections? If it is, that would prevent even admins from deleting a protected branch.</p>

**Comment by Grabauskas** [comment]
<p><a class="mention" href="/nadiajoyce">@nadiajoyce</a> I’m sure, that “Include administrators” checkbox was not checked.</p>
<p>It’s simple to test, just create a new branch protection rule and go to <a href="https://github.com/%7BOWNER%7D/%7BREPO%7D/branches" rel="nofollow noopener">https://github.com/{OWNER}/{REPO}/branches</a> link. You will see, that you can’t delete protected branches.</p>

**Comment by nadiajoyce** [comment, marked as answer]
<p>Hi <a class="mention" href="/grabauskas">@grabauskas</a> thank you for following up on this. Right you are! Sorry for any confusion. I talked with my teammates, and it’s true: protected branches cannot be deleted, even by admins. The “even admins” checkbox is more for review restrictions.</p>
<p>In order to delete a protected branch, you’d have to first remove the restrictions from it. This is because protected branches are meant to be more permanent, akin to <code>master</code>. That said, I know the OP here was asking for the ability to delete protected branches, so I will log that internally as a feature request.</p>
<p>Cheers!</p>

**Comment by hovissimo** [comment]
<p>One more vote for admins to be able to delete protected branches in sudo mode.  I understand that protected branches are ‘supposed to be’ long-lived like master, but the feature is more useful than just long-lived branches like master.</p>
<p>Our use-case:</p>
<p>My team’s workflow is that most work is done in short-lived feature branches by singly authors, but sometimes a long-lived feature branch being worked on by many authors is required.  We <em>protect</em> these long lived branches because for the force-push prevention.  We do eventually merge these branches into master and then we want to get rid of the feature branch, but we’re forced to do gymnastics and temporarily unprotect our other branches to delete it.</p>

**Comment by deanmartens** [comment]
<p>Hi, I’d like this feature request to be more extensive.  I too want to enforce that pull requests are reviewed by at least one person before being merged.  So I put that protection on all branches so this is enforced for all feature branches created over time.  However, we need to be able to remove those feature branches once they’re merged otherwise the repository is going to be a mess and using other tools to interact with branches is going to show hundreds to thousands of branches which is highly undersable.  Futhermore, if I’m the creator and thus the owner of a branch, I should be able to delete it no matter what.  Currently I find the branch protections are far too obstructive to basic operations that have nothing to do with the rules in the branch protections.  Thanks.</p>

**Comment by Startouf** [comment]
<p>Hey, we are using a similar workflow, when a lot of work is needed to shop a feature, we usually split the work and we have a hierarchy User Story (US) &gt; Feature (FT) &gt; Ticket (TK). Our US and FT are product issues and our TK issues are more tech-oriented</p>
<p>We spawn one branch per US so that we can isolate all the work that must be done to ship a new end-user feature. So we may have several branches like <code>us/42-a-user-can-report-bad-behavior</code> that can be used my multiple devs ==&gt; we add a protection to this branch.</p>
<p>Each dev will then make some individual branches <code>tk/111-open-report-behavior-popup</code> that will be merged into this <code>us/</code> branch. Then, once all the work is done, the Q/A team can do its job and once the US is validated, then we open a PR from the US branch into master, and after a successul build we merge this PR and start a release using Vincent Driessen git-flow.</p>
<p>At this stage, the US branch SHOULD disappear, but since we have protected it, we can’t do anything :’(. It would solve the problem to have a “sudo” like mode for git that would allow some people to force-remove branches (eg only a Tech lead cna merge the US into master and close this branch)</p>
<p>(I’m talking about the <code>master</code> branch, but understand the <code>develop</code> branch if referring to Vincent Driessen git flow</p>

**Comment by Winsoungnin** [comment]
<p>p=nosuchUser19890807</p>
